bayes_kit/drghmc.py

```diff
@@ -0,0 +1,437 @@
+                    raise ValueError(f"steps must be positive, but found {s} steps")
+                if type(s) is not int:
+                    raise TypeError(
+                        f"list of steps must contain integers, but found {type(s)} steps"
+                    )
+            return steps
+
+        elif isinstance(steps, Callable):  # isinstance() allows for lambda functions
+            steps = [steps(k) for k in range(self._num_proposals)]
+            for s in steps:
+                if not s > 0:
+                    raise ValueError(f"steps must be positive, but found {s} steps")
+                if type(s) is not int:
+                    raise TypeError(
+                        f"callable must return integers number of steps, but found {type(s)} steps"
+                    )
+            return steps
+
+        else:
+            raise TypeError("steps must be None, int, list, or callable")
+
+    def __iter__(self) -> Iterator[DrawAndLogP]:
+        """Use the DrGhmcDiag sampler as an iterator.
+
+        Returns:
+            Iterator[DrawAndLogP]: Iterator of draws from DrGhmcDiag sampler
+        """
+        return self
+
+    def __next__(self) -> DrawAndLogP:
+        """Yields next draw from DrGhmcDiag iterator.
+
+        Yields:
+            DrawAndLogP: Tuple of (sample, log density)
+        """
+        return self.sample()
+
+    def joint_logp(self, theta: VectorType, rho: VectorType) -> float:
+        """Joint log density of sample (theta, rho) under the cannonical distribution.
+
+        Args:
+            theta: position
+            rho: momentum
+
+        Returns:
+            float: joint log density
+        """
+        adj: float = 0.5 * np.dot(rho, self._metric * rho)
+        return self._model.log_density(theta) - adj
+
+    def leapfrog(
+        self,
+        theta: VectorType,
+        rho: VectorType,
+        stepsize: float,
+        steps: int,
+    ) -> tuple[VectorType, VectorType]:
+        """Propose new sample (theta_prop, rho_prop) with leapfrog integrator.
+
+        Discretize and solve Hamilton's equations using the leapfrog integrator to
+        deterministically propose a new sample (theta_prop, rho_prop) from the current
+        sample (theta, rho).
+
+        Copy theta because numpy's += operator mutates the original array instead of
+        creating a new one.
+
+        Initialize rho_mid by going backwards half a step so that the first full-step
+        inside the loop brings rho_mid up to +1/2 steps.
+
+        Args:
+            theta: position
+            rho: momentum
+            stepsize: stepsize in each leapfrog step
+            steps: number of leapfrog steps
+
+        Returns:
+            tuple[VectorType, VectorType]: proposed sample
+        """
+        theta = np.array(theta, copy=True)
+        _, grad = self._model.log_density_gradient(theta)
+        rho_mid = rho - 0.5 * stepsize * np.multiply(self._metric, grad).squeeze()
```

Comment:
This is clever, but it's an extra calculation if we are only going one step, so I think it's worth doing it more efficiently.
The way to do this is 

* 1/2 momentum, 
* full position
* for () in range(steps - 1)
    - full momentum
    - full position
* 1/2 momentum

It also avoids having to save that `grad` value out of the loop---the more variables can be local, the better for ease of understanding code.

Reply:
I suspect this is following on the change to base HMC from PR #30 (https://github.com/flatironinstitute/bayes-kit/pull/30/files). I've no judgment to offer on the appropriateness, but wanted to mention in case you (@bob-carpenter) think it should be reconsidered there as well.

Reply:
@bob-carpenter which implementation do you prefer?